Vectorization for spkezr

skulumani opened the pull request:
Uses the same method as spkpos and allows for a vector of ET inputs.

I haven't tested this as I'm not exactly clear as how to set up the development environment to coexist with the installed version of SpiceyPy

AndrewAnnex commented:
for testing, you can install spiceypy in development mode by running `python setup.py develop`. After that you can run the tests by running `pytest` in the project directory or by including the path to the test file.

AndrewAnnex commented:
while I haven't vectorized any of the functions in a while, I think the one-liner style I used for the other functions like `srfxpt` would be better to do unless there is a significant performance improvement. This can be tested in ipython, but it is just not high on my todo list.

skulumani commented:
Awesome. I'll take a look at it.

Thanks for the great work

On May 27, 2017 5:38 PM, "Andrew Annex" <notifications@github.com> wrote:

> while I haven't vectorized any of the functions in a while, I think the
> one-liner style I used for the other functions like srfxpt would be
> better to do unless there is a significant performance improvement. This
> can be tested in ipython, but it is just not high on my todo list.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AndrewAnnex/SpiceyPy/pull/192#issuecomment-304478003>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBaZmsBYc-ofr0VoeQPlNJ9fKtA8Ix9ks5r-JfJgaJpZM4NodQ8>
> .
>

AndrewAnnex commented:
Hey so first of all, thanks for the PR. I ended looking into my question on the performance of that I made earlier and I made a version that is about twice as fast. Once I finalized the prototype I can post it to this pr as a code block comment that you could then update your pr to reflect or we can just merge the pr as is, as it is not a huge blocker for me. Let me know what option works for you and thanks again!

skulumani commented:
That's awesome. Feel free to add your improvements and I can add them to
this function as well as some of the other frequently used ones.

On Sat, Jun 3, 2017 at 1:16 PM, Andrew Annex <notifications@github.com>
wrote:

> Hey so first of all, thanks for the PR. I ended looking into my question
> on the performance of that I made earlier and I made a version that is
> about twice as fast. Once I finalized the prototype I can post it to this
> pr as a code block comment that you could then update your pr to reflect or
> we can just merge the pr as is, as it is not a huge blocker for me. Let me
> know what option works for you and thanks again!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AndrewAnnex/SpiceyPy/pull/192#issuecomment-305988759>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBaZqDEwmI3jt6PbdW9f28tbSuGvWYVks5sAZTSgaJpZM4NodQ8>
> .
>

AndrewAnnex commented:
Hey @skulumani 

After trying a few things I came up with this version that works about twice as fast as the implementation I wrote for spkezr. 

```python
def spkezr(targ, et, ref, abcorr, obs):
    targ = stypes.stringToCharP(targ)
    ref = stypes.stringToCharP(ref)
    abcorr = stypes.stringToCharP(abcorr)
    obs = stypes.stringToCharP(obs)
    if not hasattr(et, "__iter__"):
        et = (et)
    res = []
    for i, t in enumerate(et):
        starg = stypes.emptyDoubleVector(6)
        lt = ctypes.c_double()
        libspice.spkezr_c(targ, ctypes.c_double(t), ref, abcorr, obs, starg, ctypes.byref(lt))
        res.append((stypes.vectorToList(starg), lt.value))
    return res
```
I need to think a bit about if it is better to return the results in a single list like this or if I need to split out the results into two lists (probably I do).

skulumani commented:
@AndrewAnnex Thanks for the hint on the vectorization. I'll implement it and run some tests on my own to get a feel for the codebase. 

I can also parse the results into two lists (position/state, and lighttime). I'll also include a version of this into some of the other functions (`str2et, spkezp` etc)

skulumani commented:
@AndrewAnnex 
I modified your version slightly and it has a slight penalty in my testing over the original version you posted above. 

* The `et` should be iterable, so now it's a tuple even if it's a single element
* I split out the resulting list of tuples into two lists. These lists are then squashed by numpy to remove the redundant dimension.
* This also handles both the vectorized and scalar output cases. 

I've tried several different version of getting it vectorized and you're right that this is the fastest that I can think of. 

If you think this is a good approach I can add it to this pull request, in addition to implementing the same approach for other functions.

~~~

def spkezr(targ, et, ref, abcorr, obs):
    targ = stypes.stringToCharP(targ)
    ref = stypes.stringToCharP(ref)
    abcorr = stypes.stringToCharP(abcorr)
    obs = stypes.stringToCharP(obs)
    if not hasattr(et, "__iter__"):
        et = (et,)
    res = []
    for i, t in enumerate(et):
        starg = stypes.emptyDoubleVector(6)
        lt = ctypes.c_double()
        libspice.spkezr_c(targ, ctypes.c_double(t), ref, abcorr, obs, starg, ctypes.byref(lt))
        res.append((stypes.vectorToList(starg), lt.value))
    
    state, lt = map(list, zip(*res))

    return numpy.squeeze(state), numpy.squeeze(lt)
~~~

AndrewAnnex commented:
First of all good catch on the time variable `et=(et,)`, but IMHO it is cleaner to write `et=[et]`.

I think it is more pythonic to not return two separate lists at this point, as logically users would either unpack the values in a for loop i.e.:
```python
for state, time in spkezr(...):
    print(state, time)
```
or they could trivially index the states or times once the result as it is a numpy array. The calls to map, zip, and squeeze (once you unpack in a loop) are not necessary.

Also I realized that there was no need to reallocate the starg and lt variables which resulted in a bit of a performance gain, so the function prototype is now:
```python
def spkezr9(targ, et, ref, abcorr, obs):
    targ = stypes.stringToCharP(targ)
    ref = stypes.stringToCharP(ref)
    abcorr = stypes.stringToCharP(abcorr)
    obs = stypes.stringToCharP(obs)
    starg = stypes.emptyDoubleVector(6)
    lt = ctypes.c_double()
    if not hasattr(et, "__iter__"):
        et = [et]
    res = []
    for i, t in enumerate(et):
        libspice.spkezr_c(targ, ctypes.c_double(t), ref, abcorr, obs, starg, ctypes.byref(lt))
        res.append((stypes.vectorToList(starg), lt.value))
    return numpy.array(res)
```
I will also need to think a bit about/prototype what happens when SPICE hits an error within the loop however, I think I can look through ICY/MICE to look at what the naif chooses to do and I will follow that

skulumani commented:
> I think it is more pythonic to not return two separate lists at this point
My only comment is that the other realizations of SPICE tend to return two variables from `spkezr`. My thought was that the Python version should follow suit. 

But I'm all for being as Pythonic as possible. Let me use the latest version of the vectorization and add it to this pull request.

AndrewAnnex commented:
hey, I was wondering what the status on this pr was

skulumani commented:
Sorry dude.

I've been working on a paper recently. Let me make some progress on it this
week and try to update the PR by this weekend.

On Mon, Jul 3, 2017 at 2:10 PM, Andrew Annex <notifications@github.com>
wrote:

> hey, I was wondering what the status on this pr was
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AndrewAnnex/SpiceyPy/pull/192#issuecomment-312706766>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBaZja-6-wXSW_CntwClDgH869x_Tgyks5sKS6ugaJpZM4NodQ8>
> .
>

AndrewAnnex commented:
oh, no rush on this, we all got stuff to do 😄

skulumani commented:
So there's some trickyness with array shapes. 

Many tests and other functions assume the state is a 1d array, but our current approach with return a 2d array (`(6,) vs. (6,1)')

I think that's why I originally used the `np.squeeze` to squash the extra dimension in the scalar case but have no impact when vectorized.

AndrewAnnex commented:
Ok, I took a look at your changes and I think we are very close, I made a few changes that clean it up and add error checking within the loop, also the `ltime` variable needs to be changed as there is a function with the same name. Thanks for adding a test!

```python
def spkezr(targ, et, ref, abcorr, obs):
    targ = stypes.stringToCharP(targ)
    ref = stypes.stringToCharP(ref)
    abcorr = stypes.stringToCharP(abcorr)
    obs = stypes.stringToCharP(obs)
    starg = stypes.emptyDoubleVector(6)
    lt = ctypes.c_double()
    if not hasattr(et, "__iter__"):
        et = [et]
    state = []
    times = []
    for t in et:
        libspice.spkezr_c(targ, ctypes.c_double(t), ref, abcorr, obs, starg, ctypes.byref(lt))
        checkForSpiceError(None)
        state.append(stypes.vectorToList(starg))
        times.append(lt.value)
    return numpy.squeeze(state), numpy.squeeze(times)
```

AndrewAnnex commented:
it looks like the tests broke because the phobos dsk kernel was moved to `https://naif.jpl.nasa.gov/pub/naif/generic_kernels/dsk/satellites/`, try updating that

skulumani commented:
Oh thanks. I was wondering why everything was failing but didn't put a lot
of effort to diagnosing it.

I'll update it tomorrow and test.

On Sun, Jul 9, 2017 at 9:14 PM, Andrew Annex <notifications@github.com>
wrote:

> it looks like the tests broke because the phobos dsk kernel was moved to
> https://naif.jpl.nasa.gov/pub/naif/generic_kernels/dsk/satellites/, try
> updating that
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AndrewAnnex/SpiceyPy/pull/192#issuecomment-313978880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBaZg3luALUTC3t8CxqqbMCz9IvzYpqks5sMXsQgaJpZM4NodQ8>
> .
>

AndrewAnnex commented:
hey I ended up putting in the fix for the dsk kernel for a thing I was changing last night, If you merge with the master branch again the tests should pass.

skulumani commented:
Thanks for the fix. I'll give it a shot. 

Do you have other suggestions for functions that should be vectorized? I can add a similar format to those as well. I spent a little time looking through the Matlab docs to try and see which ones were setup to handle vector inputs.

AndrewAnnex commented:
I think if we cover the ICY/MICE vectorized functions that would be enough, but we don't have to get everything done in this PR, so once your changes are up and the tests pass I will merge. I made a list a while ago for the ICY vectorized functions in the wiki https://github.com/AndrewAnnex/SpiceyPy/wiki/ICY-Vectorized-Functions, which may need to be updated.

skulumani commented:
I merged your modifications from master and Seemed to work well on my computer. 

Once this one is merged I'd be happy to start working vectorizing some of the other functions.
